### ISSUE TITLE
Tenacity @retry breaks async Dependency

### ISSUE THREAD

**Issue by rotten** [tiangolo/fastapi#3445]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

**requirements.txt**
```
fastapi==0.65.2
tenacity==7.0.0
uvicorn==0.14.0
```

**main.py**
```Python
from fastapi import Depends, FastAPI
from tenacity import retry, stop_after_attempt, wait_fixed

app = FastAPI()

@retry(stop=stop_after_attempt(4), wait=wait_fixed(10))
async def get_five():
    return 5

@app.get("/")
async def read_root(five: int = Depends(get_five)):
    return {"Hello": f"World {five}"}
```

### Description

When using `Depends` in an endpoint, if you wrap the depending async function with tenacity's `retry()` decorator, the GET will fail with a 422.  It seems to work with a regular synchronous function, and `@retry` seems to work fine with asynchronous functions in general.

If you comment out the `@retry` line, the example works fine.

### Environment

* OS: Linux - Ubuntu 20.04.2
* FastAPI Version [e.g. 0.3.0]:  0.65.2
* Python version:  3.8.5

### Additional context

uncommented:
```$  http http://localhost:8000
HTTP/1.1 422 Unprocessable Entity
content-length: 168
content-type: application/json
date: Fri, 02 Jul 2021 00:09:16 GMT
server: uvicorn

{
    "detail": [
        {
            "loc": [
                "query",
                "args"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        },
        {
            "loc": [
                "query",
                "kwargs"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        }
    ]
}
```

commented:
```
$  http http://localhost:8000
HTTP/1.1 200 OK
content-length: 19
content-type: application/json
date: Fri, 02 Jul 2021 00:09:40 GMT
server: uvicorn

{
    "Hello": "World 5"
}
```

**Comment by Mause** [comment]
This is an issue with tenacity not propagating the type information, not with fastapi

**Comment by rotten** [comment]
I was just figuring that out.  :-)   I tried making a wrapper like this:
```
import functools

from fastapi import Depends, FastAPI
from tenacity import retry, stop_after_attempt, wait_fixed

app = FastAPI()

def add_two(myfunc):
    @functools.wraps(myfunc)
    async def wrapper(*args, **kwargs):
        return await myfunc(*args, **kwargs) + 2
    return wrapper

# @retry(stop=stop_after_attempt(4), wait=wait_fixed(10))
@add_two
async def get_five():
    return 5

@app.get("/")
async def read_root(five: int = Depends(get_five)):
    return {"Hello": f"World {five}"}
```
And that works ok.

**Comment by rotten** [comment]
This was a decidedly non trivial issue to chase down in an api with many endpoints and many nested Depends.  The default return message is rather obtuse, and nothing gets logged to the fastapi console either.  I opened an issue with Tenacity and referenced this one. -- https://github.com/jd/tenacity/issues/318

**Comment by AIGeneratedUsername** [comment]
@rotten It would be nice to rename the issue or create a new one: "FastAPI shows a misleading message when decorated 'Depends' does not pass further the type information".

@Mause What do you think? Any ways to improve the message or to detect that types were not passed further in the 'Depends'?

**Comment by rotten** [comment]
Fair enough.  Now that I'm pretty sure it isn't actually a bug, updating the title makes sense.  It was _really hard_ to chase down what was happening based on the error I was getting.
